From a37ecf869f4772b1ea3b578b687b5b83dd4bd067 Mon Sep 17 00:00:00 2001
From: Alexandre Rostovtsev <tetromino@gmail.com>
Date: Fri, 27 May 2011 01:10:24 -0400
Subject: [PATCH] Prevent segfault in gtk_icon_info_load_symbolic when icon is not found (gnome bug 651209)

Users of Gentoo and Fedora have reported segfaults in nautilus-3.0.x
in gtk_icon_info_load_symbolic (see Gentoo bug 365779, Fedora bug
690357). One of them was helpful enough to post a backtrace with
debugging information: https://bugzilla.redhat.com/attachment.cgi?id=489928

The backtrace shows that the crash occurs because we are passing a
NULL icon_info to gtk_icon_info_load_symbolic in lookup_and_color_symbolic_find
in nautilus-search-bar.c; indeed, icon_info will be NULL if
gtk_icon_theme_lookup_icon fails to find an icon.

An identical problem exists in get_eject_icon in nautilus-places-sidebar.c,
where gtk_icon_info_load_symbolic_for_context and
gtk_icon_theme_lookup_by_gicon are used.

The solution is to check for the NULL icon_info, and return a
NULL GdkPixbuf. Note that returning a NULL from
lookup_and_color_symbolic_find and get_eject_icon is
safe: in both cases, the return value is only used in functions
that can safely deal with a NULL.
---
 src/nautilus-places-sidebar.c |   11 ++++++-----
 src/nautilus-search-bar.c     |    9 +++++----
 2 files changed, 11 insertions(+), 9 deletions(-)

diff --git a/src/nautilus-places-sidebar.c b/src/nautilus-places-sidebar.c
index aa08f14..de53cd1 100644
--- a/src/nautilus-places-sidebar.c
+++ b/src/nautilus-places-sidebar.c
@@ -214,7 +214,7 @@ static GdkPixbuf *
 get_eject_icon (NautilusPlacesSidebar *sidebar,
 		gboolean highlighted)
 {
-	GdkPixbuf *eject;
+	GdkPixbuf *eject = NULL;
 	GtkIconInfo *icon_info;
 	GIcon *icon;
 	int icon_size;
@@ -227,10 +227,11 @@ get_eject_icon (NautilusPlacesSidebar *sidebar,
 	icon_info = gtk_icon_theme_lookup_by_gicon (icon_theme, icon, icon_size, 0);
 
 	style = gtk_widget_get_style_context (GTK_WIDGET (sidebar));
-	eject = gtk_icon_info_load_symbolic_for_context (icon_info,
-							 style,
-							 NULL,
-							 NULL);
+	if (icon_info != NULL)
+		eject = gtk_icon_info_load_symbolic_for_context (icon_info,
+								 style,
+								 NULL,
+								 NULL);
 
 	if (highlighted) {
 		GdkPixbuf *high;
diff --git a/src/nautilus-search-bar.c b/src/nautilus-search-bar.c
index 6148807..17412d9 100644
--- a/src/nautilus-search-bar.c
+++ b/src/nautilus-search-bar.c
@@ -160,7 +160,7 @@ lookup_and_color_symbolic_find (NautilusSearchBar *bar)
 {
 	GtkIconInfo *icon_info;
 	GdkRGBA color;
-	GdkPixbuf *icon;
+	GdkPixbuf *icon = NULL;
 	GtkStyleContext *context;
 
 	context = gtk_widget_get_style_context (GTK_WIDGET (bar));
@@ -172,9 +172,10 @@ lookup_and_color_symbolic_find (NautilusSearchBar *bar)
 						"edit-find-symbolic",
 						nautilus_get_icon_size_for_stock_size (GTK_ICON_SIZE_MENU),
 						GTK_ICON_LOOKUP_GENERIC_FALLBACK);
-	icon = gtk_icon_info_load_symbolic (icon_info, &color,
-					    NULL, NULL, NULL,
-					    NULL, NULL);
+	if (icon_info != NULL)
+		icon = gtk_icon_info_load_symbolic (icon_info, &color,
+						    NULL, NULL, NULL,
+						    NULL, NULL);
 
 	gtk_style_context_restore (context);
 
-- 
1.7.5.rc3

